### DISCUSSION TITLE
CSP advice

### DISCUSSION THREAD

**Discussion by lookbusy1344** [tauri-apps/tauri#5723] [Q&A] [answered]
Can I have some advice for a basic CSP setting that will allow no external resources? I tried

`"csp": "default-src 'self'" `

but this seems to stop Material UI displaying properly in Tauri release mode. Im using MUI from local npm packages, including fonts and css.

**Comment by FabianLars** [comment, marked as answer]
`"csp": "default-src 'self'" ` would be really the best case scenario. Unfortunately this breaks many ui frameworks (especially ones for React) as you already noticed.

To be sure to only allow the absolute minimum, i'd suggest trying out debug builds (`tauri build --debug` output in `target/debug/`) and check the devtools console for errors. It's probably something like `'unsafe-eval'`/`'unsafe-inline'` for script-src/style-src + this setting: https://tauri.app/v1/api/config#securityconfig.dangerousdisableassetcspmodification

**Reply by lookbusy1344** [reply]
Thanks for the help. Everything seems to work in debug builds, as if the CSP setting wasn’t being applied, so I’m struggling to diagnose the exact problem. It would be really useful if someone could provide a few Tauri-specific CSP examples. I’ll check the main repo, maybe that contains some.

**Reply by FabianLars** [reply]
`tauri build --debug` is something different than `tauri dev` btw!
but we can go a step further. If you add the `"devtools"` feature flag to the tauri dependency in Cargo.toml the devtools will be enabled for release builds (`tauri build`) too.

**Reply by lookbusy1344** [reply]
Ah great tip about `tauri build` != `tauri dev`! For anyone else reading this, its like a half-way house between dev builds and production:

`cargo tauri build -d -b none`

This lets me use the dev tools and actually manifests the issue, unlike `cargo tauri dev`. 
I tried lots of CSP settings, all without success. Then I noticed the debug console was actually telling me the sha256 it required to work! My final, fully functional, CSP setting is:

`"csp": "default-src 'self'; style-src 'sha256-47DEQpj8HBSa+/TIXX+5JCeuBeRuj5NWZG3hSuFU='"`

Thanks again for your assistance!
